### PR TITLE
Since thindev ids are inconveniently special, give them a type

### PR DESCRIPTION
**Pull request by mulkieran**
The type can be amended or improved later, but at least it gives something
close to the maximally random number now.

Signed-off-by: mulhern <amulhern@redhat.com>

### CODE REVIEW

**Comment by mulkieran** [comment]
@agrover @trgill review?

**Comment by mulkieran** [comment]
It would be a convenience if this were merged and a new version were made today.

**Comment by agrover** [comment]
We need to have a way to make a `ThinDevId` from a u32 for calling `setup`.

Also, I have doubts about the value of a way to generate random thin_ids in this crate, when the most important thing is that they not conflict with already used values, and the crate user is the one in position to ensure that.

**Comment by mulkieran** [comment]
We can make ThinDevId serializable, which will allow us to use the type in our definition of the ThinPoolDevSave struct. That will allow us to make the values from u32s in the only place that I am aware of that we need to. I've done that part. We can implement From if we have to.

We can implement Hash and Eq, so that the client can, w/out difficulty, compare and make sure that no duplicates, etc. I've done that, also.

**Comment by mulkieran** [comment]
The test, however, is completely wrong and should be fixed.

**Comment by mulkieran** [comment]
rebase.

**Comment by mulkieran** [comment]
It is possible that what devicemapper really needs is a trait, which stratisd can implement...

**Comment by agrover** [comment]
Like I said in #67, I like this approach better but we just need a way for the client to be able to make a ThinDevId from a given number, rather than using rand.

**Comment by mulkieran** [comment]
@agrover please take a look. I started to think that simply returning a result for new method made more sense, so I gave that a try.

File: src/thindev.rs [hunk 1]
```diff
@@ -14,10 +16,59 @@ use types::TargetLine;
 
 use types::Sectors;
 
+const THIN_DEV_ID_LIMIT: u64 = 16777216; // 2 ^ 24
```
Comment (agrover) [inline src/thindev.rs]: better to express in hex? 0xFFFFFF? Or, 0x1000000 I guess, if it's being used as the first not-possible value.

File: src/lib.rs [hunk 2]
```diff
@@ -60,6 +60,8 @@ extern crate bitflags;
 #[macro_use]
 extern crate log;
 
+extern crate rand;
```
Comment (agrover) [inline src/lib.rs]: no longer needed

File: src/thindev.rs [hunk 3]
```diff
@@ -14,10 +16,59 @@ use types::TargetLine;
...
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+/// A thindev id is a 24 bit number, i.e., not a power of 2.
```
Comment (agrover) [inline src/thindev.rs]: maybe say "bit width not a power of 2"

File: src/thindev.rs [hunk 4]
```diff
@@ -14,10 +16,59 @@ use types::TargetLine;
...
+pub struct ThinDevId {
+    value: u32,
+}
+
+impl ThinDevId {
+    /// Make a new ThinDevId.
+    /// Return an error if value is too large to represent in 24 bits.
+    pub fn new_u64(value: u64) -> DmResult<ThinDevId> {
```
Comment (agrover) [inline src/thindev.rs]: super-nitpick: not just `new`? `_u64` seems redundant

Comment (agrover) [inline src/thindev.rs]: DmResult, ok.

Comment (mulkieran) [inline src/thindev.rs]: I'ld rather leave things open to new from other categories of values, hence the ```_u64```.

File: src/lib.rs (hunk 2 above)
Comment (mulkieran) [inline src/lib.rs]: Whoops!

**Pull request merged**